### PR TITLE
[SE-0046] Implementation for consistent func labels

### PR DESCRIPTION
**Pull request by manavgabhawala**
#### What's in this pull request?

Implements [SE-0046](https://github.com/apple/swift-evolution/blob/master/proposals/0046-first-label.md) to provide consistent function labels (i.e. the first parameter no longer has special behavior for function labels). 

I separated out the changes into separate commits depending on the part being changed so that its easier to review since this is a massive change across the entire repo. However, if it is preferred that this is merged as a single commit, I would be happy to squash into one commit.

I updated most of the swift code using Regex find and replace (and then manual verification where necessary) and since a lot of other repos are going to suffer source code breaking changes to make it easier for them, here are the regex strings I used:
For regular (non-generic) functions I did a search for  `func ([_a-zA-Z0-9]+?)(\s*?)\((\s*?)([a-zA-Z0-9]+?)(\s*?):` and replaced with `func \1\2(\3_ \4\5:` By estimation this worked about 95% of the time and almost never found any false positives.
For generic functions I did a second search using: `func ([_a-zA-Z0-9]+?)(\s*?)(<[\s\S]*?>)(\s*?)\((\s*?)([a-zA-Z0-9]+?)(\s*?):` and replaced with `func \1\2\3\4(\5_ \6\7:` This worked about 90% of the time but be warned it does tend to find false positives (rarely).
#### Resolved bug number:
- [SR-961](https://bugs.swift.org/browse/SR-961)
- rdar://problem/25279084

---

Before merging this pull request to apple/swift repository:
- [ ] Review changes
- [ ] Other swift repositories should be ready for a source code breaking change
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

 **Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| OS X platform | @swift-ci Please test OS X platform |
| Linux platform | @swift-ci Please test Linux platform |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

File: stdlib/private/StdlibCollectionUnittest/LoggingWrappers.swift.gyb [hunk 1]
```diff
@@ -139,16 +139,16 @@ public struct LoggingRangeReplaceableCollection<
     return base[position]
   }
 
-  public mutating func replaceSubrange<
+  public mutating func replace<
     C : Collection where C.Iterator.Element == Base.Iterator.Element
   >(
-    bounds: Range<Base.Index>, with newElements: C
+    subRange bounds: Range<Base.Index>, with newElements: C
```
Comment (gribozavr) [inline stdlib/private/StdlibCollectionUnittest/LoggingWrappers.swift.gyb]: Please don't change the standard library API.  It was already audited and it conforms to the guidelines.

**Comment by gribozavr** [comment]
CC @harlanhaskins

**Comment by gribozavr** [comment]
@manavgabhawala Thank you!  I am extremely concerned about false positives.  95% is good, but not perfect.  We can't allow any unintended changes in the standard library API.  Did you manually audit all of the changes to the `stdlib` directory?

**Comment by harlanhaskins** [comment]
@manavgabhawala Great job! I definitely share @gribozavr's concerns about API -- it's tedious but you'll need to manually verify your regex didn't cause any unintended API changes.

**Comment by manavgabhawala** [comment]
@gribozavr @harlanhaskins I did not use the regex for any of the stdlib changes for exactly that reason. I did not change any of the APIs other than the replaceSubrange function (and not just in String, in all the containers), is because almost all of the declarations were of the form replaceSubrange(_ subRange:,with:) so I thought this might be an easier change to do. However if you want, I can go ahead and change all of these back to the original form.

**Comment by harlanhaskins** [comment]
I think it's probably best to change it back for this pull request.

File: stdlib/public/core/Reflection.swift [hunk 2]
```diff
@@ -497,6 +497,11 @@ struct _EnumMirror : _Mirror {
   subscript(i: Int) -> (String, _Mirror) {
     return _getEnumChild(i, data)
   }
+<<<<<<< dde2aff82c1da3ea293c958931b1a2a7227e86db
```
Comment (harlanhaskins) [inline stdlib/public/core/Reflection.swift]: There are some conflict markers scattered through this file

**Comment by harlanhaskins** [comment]
There are some conflict markers scattered throughout `stdlib/core/Reflection.swift`

**Comment by manavgabhawala** [comment]
@harlanhaskins will fix, thanks for the catch

**Comment by manavgabhawala** [comment]
@harlanhaskins fixed all the changes requested

**Comment by harlanhaskins** [comment]
Were those conflicts from abfecfd? Thanks for fixing those so quickly!

**Comment by manavgabhawala** [comment]
No I think they were from `[9ce9784180be8ebc7744a7a295c6006f543da344](https://github.com/apple/swift/commit/9ce9784180be8ebc7744a7a295c6006f543da344)`. I had fixed them but I think I forgot to save the file before I rebased. But it should be fixed now. There are always conflicts every time there are new commits since this PR changes a lot of files.

Yeah absolutely, please make sure nothing stands out.

**Comment by harlanhaskins** [comment]
We're also going to need to have branches ready to go in SPM, corelibs, lldb, and xctest, because as soon as Top-of-Tree `swiftc` receives this change, all subprojects will fail to build.

I'll coordinate with @mxcl, @parkera, @ddunbar, and @eschaton to get their projects ready.

**Comment by manavgabhawala** [comment]
Great, sounds good. Let me know if there's anything I can help with

File: stdlib/public/core/ArrayBufferProtocol.swift [hunk 3]
```diff
@@ -69,8 +69,8 @@ public protocol _ArrayBufferProtocol : MutableCollection {
   ///
   /// - Precondition: This buffer is backed by a uniquely-referenced
   /// `_ContiguousArrayBuffer`.
-  mutating func replace<C : Collection where C.Iterator.Element == Element>(
-    subRange subRange: Range<Int>,
+  mutating func replaceSubrange<C : Collection where C.Iterator.Element == Element>(
```
Comment (harlanhaskins) [inline stdlib/public/core/ArrayBufferProtocol.swift]: Lingering API change here

File: stdlib/public/core/ArrayBufferProtocol.swift [hunk 4]
```diff
@@ -134,10 +134,10 @@ extension _ArrayBufferProtocol {
     return subscriptBaseAddress
   }
 
-  public mutating func replace<
+  public mutating func replaceSubrange<
```
Comment (harlanhaskins) [inline stdlib/public/core/ArrayBufferProtocol.swift]: And here

File: docs/SequencesAndCollections.rst [hunk 5]
```diff
@@ -306,10 +306,10 @@ range of elements, denoted by two indices, by elements from a collection with a
 ::
 
   public protocol RangeReplaceableCollectionType : MutableCollectionType {
-    mutating func replaceSubrange<
```
Comment (harlanhaskins) [inline docs/SequencesAndCollections.rst]: Another here

File: docs/StdlibAPIGuidelines.rst [hunk 6]
```diff
@@ -88,7 +88,7 @@ Subsequent Parameters
 
   .. parsed-literal::
 
-    x.replaceSubrange(r, **with:** someElements)
+	x.replace(subRange: r, **with:** someElements)
```
Comment (harlanhaskins) [inline docs/StdlibAPIGuidelines.rst]: And here

File: docs/archive/LangRefNew.rst [hunk 7]
```diff
@@ -1348,7 +1348,7 @@ otherwise empty cases in switch statements.
 
 ::
 
-  func classifyPoint(point: (Int, Int)) {
```
Comment (harlanhaskins) [inline docs/archive/LangRefNew.rst]: API change

File: docs/archive/Objective-CInteroperability.rst [hunk 8]
```diff
@@ -496,21 +496,21 @@ Overloading
   name, unless using the "selector syntax" for defining methods::
 
     // 1. foo:baz:
-    func foo(Int bar, Int baz);
+    func foo(_ bar: Int, baz: Int)
```
Comment (harlanhaskins) [inline docs/archive/Objective-CInteroperability.rst]: I applaud you for finding and fixing these.

@gribozavr What do you think about these changes? It's an archived document with old syntax.

File: docs/archive/Objective-CInteroperability.rst [hunk 9]
```diff
@@ -496,21 +496,21 @@ Overloading
...
   
     // 2. foo:qux:
-    func foo(Int bar, Int qux);
-  
+    func foo(_ bar: Int bar, qux: Int)
```
Comment (harlanhaskins) [inline docs/archive/Objective-CInteroperability.rst]: `Int bar` repeated here

File: docs/proposals/OptimizerEffects.rst [hunk 10]
```diff
@@ -185,17 +185,17 @@ state.
       var storage: ArrayStorage
 
       @preserve_unique
-      mutating func replaceSubrange<
```
Comment (harlanhaskins) [inline docs/proposals/OptimizerEffects.rst]: API change

Comment (harlanhaskins) [inline docs/proposals/OptimizerEffects.rst]: (Even though these are just docs)

File: docs/archive/LangRefNew.rst (hunk 7 above)
Comment (gribozavr) [inline docs/archive/LangRefNew.rst]: Is it?

File: docs/archive/Objective-CInteroperability.rst (hunk 8 above)
Comment (gribozavr) [inline docs/archive/Objective-CInteroperability.rst]: Agreed -- I'd rather avoid making further edits to this document.

File: docs/proposals/ValueSemantics.rst [hunk 11]
```diff
@@ -179,7 +179,7 @@ Here's a version of cycle_length that works when state is a mutable
 value type::
 
  func cycle_length<State>(
-   s : State, mutate : ( [inout] State ) -> () 
+   + s : State, mutate : ( [inout] State ) -> ()
```
Comment (harlanhaskins) [inline docs/proposals/ValueSemantics.rst]: `+` instead of `_`

File: docs/archive/LangRefNew.rst (hunk 7 above)
Comment (harlanhaskins) [inline docs/archive/LangRefNew.rst]: It's not actually a change in API, but I don't think changing the doc for API naming is in scope for this change.

Though I'm not against this change if you wanna just take it.

File: stdlib/private/StdlibCollectionUnittest/CheckRangeReplaceableCollectionType.swift [hunk 12]
```diff
@@ -75,7 +75,7 @@ internal struct ReplaceRangeTest {
     self.newElements = newElements.map(OpaqueValue.init)
     self.rangeSelection = rangeSelection
     self.expected = expected
-    self.loc = SourceLoc(file, line, comment: "replaceSubrange() test data")
```
Comment (harlanhaskins) [inline stdlib/private/StdlibCollectionUnittest/CheckRangeReplaceableCollectionType.swift]: replaceSubrange -> replace

File: stdlib/private/StdlibCollectionUnittest/MinimalCollections.swift.gyb [hunk 13]
```diff
@@ -1134,11 +1134,11 @@ public struct ${Self}<T> : RangeReplaceableCollection {
   public mutating func replaceSubrange<
     C : Collection where C.Iterator.Element == T
   >(
-    subRange: Range<${Index}>, with newElements: C
+    _ subRange: Range<${Index}>, with newElements: C
   ) {
     let oldCount = count
     elements.replaceSubrange(
-      subRange.startIndex.position..<subRange.endIndex.position,
+	  subRange.startIndex.position..<subRange.endIndex.position,
```
Comment (harlanhaskins) [inline stdlib/private/StdlibCollectionUnittest/MinimalCollections.swift.gyb]: nit-pick: Indentation

File: docs/archive/Objective-CInteroperability.rst (hunk 8 above)
Comment (manavgabhawala) [inline docs/archive/Objective-CInteroperability.rst]: @gribozavr so should I get rid of these changes then, or leave them in?

**Comment by harlanhaskins** [comment]
I can't see the file right now, but whilst running tests I found a conflict marker in `test/expr/unary/selector/selector.swift`

File: docs/archive/Objective-CInteroperability.rst (hunk 8 above)
Comment (gribozavr) [inline docs/archive/Objective-CInteroperability.rst]: Let's revert the changes from this file.

File: stdlib/public/core/ArrayBuffer.swift [hunk 14]
```diff
@@ -392,8 +391,8 @@ extension _ArrayBuffer {
       }
       else {
         var refCopy = self
-        refCopy.replace(
-          subRange: i...i, with: 1, elementsOf: CollectionOfOne(newValue))
+        refCopy.replaceSubrange(
```
Comment (harlanhaskins) [inline stdlib/public/core/ArrayBuffer.swift]: API change here

**Comment by harlanhaskins** [comment]
Notes for remaining files (GitHub's diff tool can't handle this right now)

### Standard Library

- `stdlib/core/SliceBuffer.swift`
  - `replace` -> `replaceSubrange` (line 80)
  - `replace` -> `replaceSubrange` (line 103)
  - `replace` -> `replaceSubrange` (line 152)
- `stdlib/core/String.swift`
  - `replaceSubrange` argument name changed (line 686)
- `stdlib/core/StringCharacterView.swift`
  - `replaceSubrange` argument name changed (line 260)

**Comment by manavgabhawala** [comment]
@harlanhaskins I fixed all of these so far. I'm rerunning all the tests before I push the new commit though (just to be sure I didn't accidentally break anything). 
Of course :) 
Don't worry, I'm glad you are nit-picky, its absolutely necessary to keep this massive code base in order!

**Comment by manavgabhawala** [comment]
@harlanhaskins I just fixed all the changes requested so far, I don't think I did but, hopefully I didn't forget any.

**Comment by manavgabhawala** [comment]
@harlanhaskins fixed `classifyPoint`

**Comment by harlanhaskins** [comment]
Small change; almost positively inconsequential, the changes in `test/DebugInfo/AlignedAt.swift` change around some of the arguments, which changes the mangled name. You updated the test for it, but it might be best just to change the declarations.

@gribozavr Thoughts?

**Comment by manavgabhawala** [comment]
In some places I changed the declaration and in others I changed the mangled names to test for both scenarios. But if you think it's better to just change the declaration I can fix that

**Comment by harlanhaskins** [comment]
That'd be best. Also thanks for changing the replace/replaceSubrange things. Those look good now 👍

**Comment by harlanhaskins** [comment]
Hmm...what happened with `test/NameBinding/library.swift`? Were these tests broken before the change?

**Comment by manavgabhawala** [comment]
For `test/NameBinding/library.swift`, it used to give us that 'wrong' and weird diagnostic, but with these changes they no longer happen, so I updated the test case to reflect that. Should I have done something else?

**Comment by harlanhaskins** [comment]
Nope, I think that's the correct thing to do. @gribozavr? I'd like to know why the same warning didn't show up since we just changed the declaration such that it would parse the same as before.

**Comment by manavgabhawala** [comment]
Also, what are you sure you meant `test/DebugInfo/AlignedAt.swift` I couldn't find any test case like that?

**Comment by harlanhaskins** [comment]
Oops, sorry, I meant `test/DebugInfo/inlinedAt.swift`

**Comment by gribozavr** [comment]
@manavgabhawala I think you need to adjust `test/NameBinding/Inputs/imported_module.swift` to revert to old diagnostics in `test/NameBinding/library.swift`.  This is important since with this change the `over1` functions no longer have the same labels.

**Comment by manavgabhawala** [comment]
@gribozavr @harlanhaskins fixed both `test/DebugInfo/inlinedAt.swift` and `test/NameBinding/library.swift `

**Comment by manavgabhawala** [comment]
@harlanhaskins I think you meant `test/SILGen/coverage_closures.swift` not `test/SILGen/coverage_exceptions.swift`, I fixed the closures one.
Also `test/expr/expressions.swift` there's no change on line 758, you probably meant another file or a different line? 
Fixed all the others, except for `test/SourceKit/DocSupport/doc_swift_module.swift.response` and `validation-test/SIL/crashers/017-swift-decl-walk.sil`, because I'm not sure what's wrong or what to fix.

**Comment by harlanhaskins** [comment]
For `test/expr/expressions.swift`: sorry, it was 798.
I don't know what to do for `doc_swift_module.swift.response`, but my instinct is to remove the diagnostic test from this commit and commit it later once we know what to do.

**Comment by manavgabhawala** [comment]
I fixed, `test/expr/expressions.swift`, and pushed a new commit but github is being weird and not showing the new commit for some reason and when I `git push` it says everything up to date. For `doc_swift_module.swift.response `, if I revert this test case to the original, it will fail.

**Comment by gribozavr** [comment]
If `validation-test/SIL/crashers/017-swift-decl-walk.sil` does not crash anymore, then it is an improvement!  Please adjust the RUN lines so that we expect a clean exit and move the test to `validation-test/SIL/crashers_fixed/`.

**Comment by harlanhaskins** [comment]
@nkcsgexi Any ideas how @manavgabhawala should add a test for a new diagnostic message in `test/SourceKit/DocSupport/doc_swift_module.swift.response`?

**Comment by nkcsgexi** [comment]
The test failure shows us two files are different  doc_swift_module.swift.response and doc_swift_module.swift.response.tmp. What I typically do is (1) manually verify the diff is reflects desirable outcomes, and (2) update  doc_swift_module.swift.response to *.tmp, so that they are identical.

**Comment by harlanhaskins** [comment]
I think the problem is that his diagnostic includes some absolute file paths, which fails on anyone else's machine...

**Comment by nkcsgexi** [comment]
oh. Do we need to add an absolute file path into diagnosis? I think file names will suffice: e.g.
 StringRef FileName = llvm::sys::path::filename(FullPath); // Give you the file name from an absolute path.

**Comment by harlanhaskins** [comment]
Sorry, I wasn't specific. This PR isn't generating filenames. What's happening is @manavgabhawala's `doc_swift_module.swift.response` file is generated by the existing code with extra bits appended to the end, specifically:

```
[
  {
    key.line: 81,
    key.column: 17,
    key.filepath: "/Users/manavgabhawala/Documents/Xcode/Swift/swift/test/SourceKit/DocSupport/Inputs/main.swift",
    key.severity: source.diagnostic.severity.error,
    key.description: "method 'protMeth(a:)' has different argument names from those required by protocol 'Prot' ('protMeth')",
    key.fixits: [
      {
        key.offset: 1092,
        key.length: 0,
        key.sourcetext: "_ "
      }
    ],
    key.diagnostics: [
      {
        key.line: 35,
        key.column: 8,
        key.filepath: "/Users/manavgabhawala/Documents/Xcode/Swift/swift/test/SourceKit/DocSupport/Inputs/main.swift",
        key.severity: source.diagnostic.severity.note,
        key.description: "requirement 'protMeth' declared here"
      }
    ]
  }
]
```

and we don't know why.

**Comment by harlanhaskins** [comment]
Also, @manavgabhawala `test/SILOptimizer/array_mutable_assertonly.swift` has a `func use(a:` that needs to be changed.

**Comment by harlanhaskins** [comment]
Wait, no, I see the issue here. @manavgabhawala You need to change all of the `protMeth(a:)`'s inside `test/SourceKit/DocSupport/Inputs/main.swift`. `class BC2` and `class SubC2` both have incorrect labels.

Sorry to bother you, @nkcsgexi.

**Comment by nkcsgexi** [comment]
I see. Thanks for the explanation! When generating doc_swift_module.swift.response, one can assume that the module compiles without any warning message. So the missing part is to update test/SourceKit/DocSupport/Inputs/main.swift and test/SourceKit/DocSupport/Inputs/cake.swift so that no diagnosis is generated.

**Comment by harlanhaskins** [comment]
Also @manavgabhawala (I keep finding more things, running the full test suite on OS X)
- `test/1_stdlib/UIKit.swift` needs updating
  - `printDevice(o:)` -> `printDevice(_ o:)`
  - `printInterface(o:)` -> `printInterface(_ o:)`

**Comment by manavgabhawala** [comment]
I ran and tested all with `utils/build-script -R -T` Should I test with a different command? Also, the `test/SILOptimizer/array_mutable_assertonly.swift ` was unsupported for me for some reason that's why I didn't catch it. Fixed it though. 
Also updated `test/SourceKit/DocSupport/Inputs/main.swift`, not sure if I need to change `doc_swift_module.swift.response` at all or not? But the test case does pass when I run it.
Also fixed `UIKit.swift`, which was unsupported when I ran it.

**Comment by manavgabhawala** [comment]
As for `validation-test/SIL/crashers/017-swift-decl-walk.sil `, I moved it to the right directory and removed the `--crash` from the file.

**Comment by gribozavr** [comment]
@swift-ci Please test

**Comment by harlanhaskins** [comment]
@gribozavr Oh no; we can't test this without the changes in SPM...

**Comment by harlanhaskins** [comment]
@manavgabhawala Also it looks like the extra diagnostics didn't get deleted from `doc_swift_module.swift.response`.

**Comment by gribozavr** [comment]
@harlanhaskins I know.  I hoped to get at least some CI feedback.

**Comment by gribozavr** [comment]
And we have some:

```
/Users/buildnode/jenkins/workspace/swift-PR-osx/swift/stdlib/public/core/ObjCMirrors.swift:46:41: error: missing argument label 'object:' in call
    return _getClassPlaygroundQuickLook(object)
```

**Comment by harlanhaskins** [comment]
Interesting! I didn't get that one when I built with `--test --validation-test`...

**Comment by harlanhaskins** [comment]
Oh, these are new changes from master?

**Comment by manavgabhawala** [comment]
@harlanhaskins can I rebase now, if you are done reviewing? so that I can fixup the new extra commits that I pushed and also get all the new changes to master?

**Comment by harlanhaskins** [comment]
Yes, absolutely! Thanks again!

We're going to coordinate this with corelibs-foundation, XCTest, and SPM, to get all of these merged together at the same time.

There will undoubtedly be at least one more rebase in your future.

**Comment by shahmishal** [comment]
@swift-ci Please test

**Comment by shahmishal** [comment]
@swift-ci Please smoke test

**Comment by harlanhaskins** [comment]
Smoke test passed. Are we set to merge?

**Comment by harlanhaskins** [comment]
Some conflicts just arose 😕

@manavgabhawala Mind fixing these and squashing into one commit?

**Comment by harlanhaskins** [comment]
You just need to delete the definitions for `insetBy` and the like, inside `stdlib/core/SDK/CoreGraphics/CoreGraphics.swift`.

**Comment by manavgabhawala** [comment]
@harlanhaskins do you want just one commit total, or squash into the current 5, correctly?

**Comment by harlanhaskins** [comment]
@manavgabhawala One commit total, if possible.

**Comment by manavgabhawala** [comment]
@harlanhaskins done. Should we re- smoke test?

**Comment by harlanhaskins** [comment]
@manavgabhawala I don't think that will be necessary. @gribozavr?

**Comment by harlanhaskins** [comment]
@swift-ci Please smoke test

**Comment by harlanhaskins** [comment]
Thank you, @manavgabhawala!
